documentcloud/documents/models/document.py

```diff
@@ -445,12 +459,108 @@ def set_page_text(self, page_text_infos):
+            "[SET PAGE TEXT] upload %d - %d mb",
+            self.pk,
+            sum(len(i) for i in file_contents) / 1000 / 1000,
+        )
+        for name, contents in zip(file_names, file_contents):
+            logger.info(
+                "[SET PAGE TEXT] upload %d - %s: %d mb",
+                self.pk,
+                name,
+                len(contents) / 1000 / 1000,
+            )
         # reverse the lists to upload the larger files first
         storage.async_upload(file_names[::-1], file_contents[::-1], access=self.access)
 
         return json_text
 
+    def _set_page_positions(self, pages, file_names, file_contents):
+        """Handle grafting page positions back into the document"""
+
+        current_pdf_contents = storage.open(self.doc_path, "rb").read()
+        current_pdf = pymupdf.open(stream=current_pdf_contents)
+        start_page = pages[0]["page_number"]
+        stop_page = pages[-1]["page_number"]
+        grafted_pdf = self._init_graft_pdf(current_pdf, start_page, stop_page)
+        current_pdf.close()
+
+        for page in pages:
+            page_number = page["page_number"]
+            if page.get("positions"):
+                logger.info(
+                    "[SET PAGE TEXT] %d - positions page %d", self.pk, page_number
+                )
+                file_names.append(
+                    path.page_text_position_path(self.pk, self.slug, page_number)
+                )
+                positions = [{**p.pop("metadata", {}), **p} for p in page["positions"]]
+                file_contents.append(json.dumps(positions).encode("utf-8"))
+
+                logger.info("[SET PAGE TEXT] %d - graft page %d", self.pk, page_number)
+                # create the overlay file
+                self._graft_page(
+                    page["positions"],
+                    grafted_pdf,
+                    page_number - start_page,
+                )
+
+        # graft the overlay pages back onto the original document
+        base_pdf = Pdf.open(BytesIO(current_pdf_contents))
+        overlay_pdf = Pdf.open(BytesIO(grafted_pdf.tobytes()))
+        grafted_pdf.close()
+
+        for i in range(start_page, stop_page + 1):
+            base_page = PikePage(base_pdf.pages[i])
+            overlay_page = PikePage(overlay_pdf.pages[i - start_page])
+            base_page.add_overlay(overlay_page, Rectangle(*base_page.trimbox))
+
+        buffer = BytesIO()
+        base_pdf.save(buffer)
+        buffer.seek(0)
+        return buffer.read()
+
+    def _init_graft_pdf(self, current_pdf, start_page, stop_page):
+        """Initialize a new PDF to graft OCR text onto"""
+        grafted_pdf = pymupdf.open()
+
+        for pdf_page in current_pdf.pages(start_page, stop_page + 1):
+            grafted_pdf.new_page(
+                width=pdf_page.rect.width,
+                height=pdf_page.rect.height,
+            )
+        return grafted_pdf
+
+    def _graft_page(self, positions, grafted_pdf, page):
```

Comment:
Strong test candidate. Are the text positions inserted into the PDF with the expected coordinates?